File: frontend/src/api/curriculum.ts
import { apiFetch } from './client'

export type TrackSubject = {
  id: string
  program_id: string
  track: 'CORE' | 'CYBER' | 'AI_DS' | 'AI_ML' | string
  subject_id: string
  is_elective: boolean
  sessions_override?: number | null
  created_at: string
}

export type TrackSubjectCreate = {
  program_code: string
  academic_year_number: number
  track: 'CORE' | 'CYBER' | 'AI_DS' | 'AI_ML' | string
  subject_code: string
  is_elective: boolean
  sessions_override?: number | null
}

export async function listTrackSubjects(params: {
  program_code: string
  academic_year_number: number
}): Promise<TrackSubject[]> {
  const qs = new URLSearchParams({
    program_code: params.program_code,
    academic_year_number: String(params.academic_year_number),
  })
  return apiFetch<TrackSubject[]>(`/api/curriculum/track-subjects?${qs.toString()}`)
}

export async function createTrackSubject(payload: TrackSubjectCreate): Promise<TrackSubject> {
  return apiFetch<TrackSubject>('/api/curriculum/track-subjects', {
    method: 'POST',
    body: JSON.stringify(payload),
  })
}

export async function deleteTrackSubject(id: string): Promise<{ ok: true }> {
  return apiFetch<{ ok: true }>(`/api/curriculum/track-subjects/${id}`, { method: 'DELETE' })
}

File: backend/schemas/curriculum.py
from __future__ import annotations

import uuid
from datetime import datetime

from pydantic import BaseModel, Field


class TrackSubjectBase(BaseModel):
    program_code: str = Field(min_length=1)
    academic_year_number: int = Field(ge=1, le=4)
    track: str = Field(min_length=1)
    subject_code: str = Field(min_length=1)
    is_elective: bool = False
    sessions_override: int | None = Field(default=None, ge=0)


class TrackSubjectCreate(TrackSubjectBase):
    pass


class TrackSubjectUpdate(BaseModel):
    track: str | None = None
    is_elective: bool | None = None
    sessions_override: int | None = Field(default=None, ge=0)


class TrackSubjectOut(BaseModel):
    id: uuid.UUID
    program_id: uuid.UUID
    academic_year_id: uuid.UUID
    track: str
    subject_id: uuid.UUID
    is_elective: bool
    sessions_override: int | None
    created_at: datetime

    class Config:
        from_attributes = True
